### ISSUE TITLE
cfeeecd381f3f0fa7ee3d19f902975a77c5100e8 is causing near-100% TCP BTL timeouts in Cisco MTT

### ISSUE THREAD

**Issue by jsquyres** [open-mpi/ompi#2936]
The cfeeecd381f3f0fa7ee3d19f902975a77c5100e8 commit (and its follow on, c62498ab3dba005b270b3f1f05eb39147644918e) are causing failures in Cisco MTT using the TCP BTL.  Each Cisco server has 5 IP interfaces (one of which is localhost).

[MTT link showing massive numbers of timeouts](https://mtt.open-mpi.org/index.php?do_redir=2390).  I stopped the master TCP-based runs after that.

When I run with these two commits, this is what happens (you can see this in MTT, too):

```
$ srun hostname
mpi028
mpi029
$ mpirun --mca btl tcp,vader,self ring_c
Process 0 sending 10 to 1, tag 201 (32 processes in ring)
Process 0 sent to 1
[mpi029:13014] btl: tcp: Incoming connection from 10.2.0.28 does not match known addresses for peer [[28017,1],15]. Drop !
```

10.2.0.28 is the IP address of one of the interfaces on the mpi028 server.

**Comment by jsquyres** [comment]
I'd like to revert these two commits by tonight's build if they aren't fixed yet.  I'll file a PR.

**Comment by bosilca** [comment]
I don't have access to machines with 5 interfaces, but I cannot replicate this issue on a cluster with 3 interfaces, even if I try to force the most restrictive behaviors of the TCP BTL.

**Comment by jsquyres** [comment]
FWIW, I can get this to happen with 2 IP interfaces.  Here's using two individual subnets -- they work fine:

```
$ mpirun -np 2 --map-by node --mca btl_tcp_if_include 10.0.0.0/16 --mca btl tcp,vader,self ring_c
Process 0 sending 10 to 1, tag 201 (2 processes in ring)
Process 0 sent to 1
Process 0 decremented value: 9
Process 0 decremented value: 8
[...snip...]
Process 0 decremented value: 0
Process 0 exiting
Process 1 exiting

$ mpirun -np 2 --map-by node --mca btl_tcp_if_include 10.3.0.0/16 --mca btl tcp,vader,self ring_c
Process 0 sent to 1
Process 0 decremented value: 9
Process 0 decremented value: 8
[...snip...]
Process 0 decremented value: 0
Process 0 exiting
Process 1 exiting
```

But if I run with both of those subnets, I get the error:

```
$ mpirun -np 2 --map-by node --mca btl_tcp_if_include 10.3.0.0/16,10.0.0.0/16 --mca btl tcp,vader,self ring_c
Process 0 sending 10 to 1, tag 201 (2 processes in ring)
[mpi010:24525] btl: tcp: Incoming connection from 10.3.0.4 does not match known addresses for peer [[11453,1],0]. Drop !
```

@bosilca and I are discussing in IM.

**Comment by jsquyres** [comment]
It seems that this has something to do with a mask of `/16` -- @bosilca was able to replicate.

It is unlikely that #2937 will be necessary; @bosilca is working on it.

**Comment by bosilca** [comment]
It turns out the mask was just a symptom not the cause. The real reason was a cut of the loop over the locally known remote interface during the connection establishment. This issue has been fixed by bc2890e.
